### PR TITLE
Use Borsh ser/deser in TransactionWrapper and BitcoinHeaderWrapper

### PR DESCRIPTION
**Pull request by jfldde**
# Description

- Implement `BorshSerialize` and `BorshDeserializaiton` on `TransactionWrapper` and `BitcoinHeaderWrapper` by manually ser/deser every inner fields
- Remove usage of bitcoin consensus encoding

### CODE REVIEW

**Review by eyusufatik — COMMENTED** [review]
what gains do we get from not using consensus encode? which is very similar to borsh encode when you think about it btw.

File: crates/bitcoin-da/src/spec/header.rs [hunk 1]
```diff
@@ -71,7 +70,7 @@ impl HeaderWrapper {
 }
 
 /// BitcoinHeaderWrapper is a wrapper around BitcoinHeaderWrapper to implement borsh serde
-#[derive(Clone, PartialEq, Eq, Debug, Hash, Deserialize, Serialize)]
+#[derive(Clone, PartialEq, Eq, Debug, Hash, Serialize, Deserialize)]
 #[repr(transparent)]
 #[serde(transparent)]
 pub struct BitcoinHeaderWrapper {
```
Comment (yaziciahmet) [inline crates/bitcoin-da/src/spec/header.rs:76]: If it's just a wrapper, imo it's better to have it as `pub struct BitcoinHeaderWrapper(BitcoinHeader)`

File: crates/bitcoin-da/src/spec/header.rs [hunk 2]
```diff
@@ -71,7 +70,7 @@ impl HeaderWrapper {
 }
 
 /// BitcoinHeaderWrapper is a wrapper around BitcoinHeaderWrapper to implement borsh serde
-#[derive(Clone, PartialEq, Eq, Debug, Hash, Deserialize, Serialize)]
+#[derive(Clone, PartialEq, Eq, Debug, Hash, Serialize, Deserialize)]
```
Comment (yaziciahmet) [inline crates/bitcoin-da/src/spec/header.rs:73]: Do we still need serde derives?

File: crates/bitcoin-da/src/spec/transaction.rs [hunk 3]
```diff
@@ -11,12 +12,12 @@ use serde::{Deserialize, Serialize};
 #[repr(transparent)]
 #[serde(transparent)]
 pub struct TransactionWrapper {
```
Comment (yaziciahmet) [inline crates/bitcoin-da/src/spec/transaction.rs:14]: Same wrapper suggestion as above

**Review by yaziciahmet — CHANGES_REQUESTED** [review]

**Comment by eyusufatik** [comment]
can you do benchmarks on test_all_flow? if you don't have the right commands let me know over slack. also @yaziciahmet recently asked me those commands as well.

**Review by yaziciahmet — APPROVED** [review]

**Comment by kpp** [comment]
Actually it would be better to test borsh in isolation - only to de-serialize inputs and exit, because now it takes only a small portion of our execution (less then 1% of total cycles). @yaziciahmet did it in https://github.com/chainwayxyz/citrea/pull/943 for comparison.

**Comment by eyusufatik** [comment]
> Actually it would be better to test borsh in isolation - only to de-serialize inputs and exit, because now it takes only a small portion of our execution (less then 1% of total cycles). @yaziciahmet did it in https://github.com/chainwayxyz/citrea/pull/943 for comparison.

yes let's do it

**Comment by eyusufatik** [comment]
> can you do benchmarks on test_all_flow? if you don't have the right commands let me know over slack. also @yaziciahmet recently asked me those commands as well.

forgot these were tests on MockDA. we should wait for merge until @jfldde 's bitcoin e2e testing stuff is ready and then benchmark.

**Comment by yaziciahmet** [comment]
With the merge of #973 , new bitcoin 0.32.2 version has changed the consensus_encode mechanism, and it is no longer usable with the Borsh ser/de. We tried couple of different methods to read/write from and to Borsh to consensus encodable, but it failed. I tested this PR and it seems to solve this issue. It might be required now that we merge this PR in @eyusufatik @kpp

**Pull request merged**